pkg/dns/dns.go

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2023 The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dns
+
+import (
+	"fmt"
+	"net"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	config_cluster_v3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	"github.com/miekg/dns"
+)
+
+type DNSResolver struct {
+	DnsResolverChan   chan map[string]time.Duration
+	client            *dns.Client
+	resolvConfServers []string
+	cache             map[string]domainCacheEntry
+	sync.Mutex
+}
+
+// domainCacheEntry stores dns result with expiry time, and also response to trigger dns refresh
+type domainCacheEntry struct {
+	value  []string
+	expiry time.Time
+	timer  *time.Timer
+}
+
+func NewDNSResolver() (*DNSResolver, error) {
+	r := &DNSResolver{
+		DnsResolverChan: make(chan map[string]time.Duration),
+		cache:           map[string]domainCacheEntry{},
+		client: &dns.Client{
+			DialTimeout:  5 * time.Second,
+			ReadTimeout:  5 * time.Second,
+			WriteTimeout: 5 * time.Second,
+		},
+	}
+
+	dnsConfig, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+	if dnsConfig != nil {
+		for _, s := range dnsConfig.Servers {
+			r.resolvConfServers = append(r.resolvConfServers, net.JoinHostPort(s, dnsConfig.Port))
+		}
+	}
+
+	return r, nil
+}
+
+func (r *DNSResolver) Start() {
+	go r.startResolver()
+}
+
+// startResolver watches the DnsResolver Channel
+func (r *DNSResolver) startResolver() {
+	for domains := range r.DnsResolverChan {
+		r.resolveDomains(domains)
+	}
+}
+
+// resolveDomains takes a map of hostnames and refresh rate
+func (r *DNSResolver) resolveDomains(domains map[string]time.Duration) {
+	r.Lock()
```

Comment:
```suggestion
	r.RLock()
```
Wouldn't it be better to use a read/write lock?